src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,909 @@
+ * The patch is approximated by the control points,
+ * parametrized from u=0 to u=1 and v=0 to v=1.
+ * 
+ * Contains a 2D array of positive weights to represent a rational Bezier patch.
+ * Nonrational Bezier patches are identified by an empty weights array.
+ * Algorithms for Rational Bezier curves derived from 
+ * Gerald Farin, "Algorithms for rational Bezier curves"
+ * Computer-Aided Design, Volume 15, Number 2, 1983,
+ */
+template <typename T>
+class BezierPatch
+{
+public:
+  using PointType = Point<T, 3>;
+  using VectorType = Vector<T, 3>;
+  using PlaneType = Plane<T, 3>;
+
+  using CoordsMat = axom::Array<PointType, 2>;
+  using NumsMat = axom::Array<T, 2>;
+
+  using BoundingBoxType = BoundingBox<T, 3>;
+  using OrientedBoundingBoxType = OrientedBoundingBox<T, 3>;
+  using BezierCurveType = primal::BezierCurve<T, 3>;
+
+  AXOM_STATIC_ASSERT_MSG(
+    std::is_arithmetic<T>::value,
+    "A Bezier Patch must be defined using an arithmetic type");
+
+public:
+  /*!
+   * \brief Constructor for a Bezier Patch that reserves space for
+   *  the given order of the surface
+   *
+   * \param [in] order the order of the resulting Bezier patch
+   * \pre order is greater than or equal to -1.
+   */
+  BezierPatch(int ord_u = -1, int ord_v = -1)
+  {
+    SLIC_ASSERT(ord_u >= -1 && ord_v >= -1);
+
+    const int sz_u = utilities::max(0, ord_u + 1);
+    const int sz_v = utilities::max(0, ord_v + 1);
+
+    m_controlPoints.resize(sz_u, sz_v);
+
+    makeNonrational();
+  }
+
+  /*!
+   * \brief Constructor for a Bezier Patch from a 2D array of coordinates
+   *
+   * \param [in] pts an array with ord_u+1 arrays with ord_v+1 control points
+   * \param [in] ord_u The patches's polynomial order on the first axis
+   * \param [in] ord_v The patches's polynomial order on the second axis
+   * \pre order in both directions is greater than or equal to zero
+   *
+   * Elements of pts[k] are mapped to control nodes (p, q) lexicographically, i.e.
+   * pts[0]               -> nodes[0, 0],     ..., pts[ord_u]           -> nodes[ord_u, 0]
+   * pts[ord_u+1]         -> nodes[0, 1],     ..., pts[2*ord_u]         -> nodes[ord_u, 1]
+   *                                          ...
+   * pts[ord_v*(ord_u-1)] -> nodes[0, ord_v], ..., pts[ord_v,ord_u] -> nodes[ord_u, ord_v]
+   */
+  BezierPatch(PointType* pts, int ord_u, int ord_v)
+  {
+    SLIC_ASSERT(pts != nullptr);
+    SLIC_ASSERT(ord_u >= 0 && ord_v >= 0);
+
+    const int sz_u = utilities::max(0, ord_u + 1);
+    const int sz_v = utilities::max(0, ord_v + 1);
+
+    m_controlPoints.resize(sz_u, sz_v);
+
+    for(int t = 0; t < sz_u * sz_v; ++t) m_controlPoints.flatIndex(t) = pts[t];
+
+    makeNonrational();
+  }
+
+  /*!
+   * \brief Constructor for a Rational Bezier Path from an array of coordinates and weights
+   *
+   * \param [in] pts a matrix with (ord_u+1) * (ord_v+1) control points
+   * \param [in] weights a matrix with (ord_u+1) * (ord_v+1) positive weights
```

Comment:
Is there a particular matrix type required? For example, does passing a linear array of length (ord_u+1) * (ord_v+1) work?

Reply:
This is a good point. Currently it's expected that the `axom::Array` be a 2D matrix, since that makes the assignment simple, but it would also be practical to allow 1D `axom::Array` vectors to mirror the behavior of how it works with C-style arrays. I've added that functionality in another function, and modified several of the comments for clarity.